funcx/executors/high_throughput/worker_map.py

```diff
@@ -38,9 +51,174 @@ def remove_worker(self, worker_id):
+                new_worker_map : dict
+                   {worker_type: total_number_of_containers,...}.
+
+                Returns
+                ---------
+                List of removed worker types.
+        """
+
+        spin_downs = []
+        for worker_type in new_worker_map:
+            if new_worker_map[worker_type] < self.total_worker_type_counts[worker_type]:
+                num_remove = self.total_worker_type_counts[worker_type] - new_worker_map[worker_type]
+
+                logger.info("[WORKER_REMOVE] Removing {} workers of type {}".format(num_remove, worker_type))
+                for i in range(num_remove):
+                    spin_downs.append(worker_type)
+        return spin_downs
+
+    def add_worker(self, worker_id=str(random.random()),
+                   mode='no_container',
+                   worker_type='RAW',
+                   container_uri=None,
+                   walltime=1,
+                   address=None,
+                   debug=None,
+                   worker_port=None,
+                   logdir=None,
+                   uid=None):
+        """ Launch the appropriate worker
+
+        Parameters
+        ----------
+        worker_id : str
+           Worker identifier string
+        mode : str
+           Valid options are no_container, singularity
+        walltime : int
+           Walltime in seconds before we check status
+
+        """
+
+        debug = ' --debug' if debug else ''
+
+        worker_id = ' --worker_id {}'.format(worker_id)
+
+        cmd = (f'funcx-worker {debug}{worker_id} '
+               f'-a {address} '
+               f'-p {worker_port} '
+               f'-t {worker_type} '
+               f'--logdir={logdir}/{uid} ')
+
+        logger.info("Command string :\n {}".format(cmd))
+
+        if mode == 'no_container':
+            modded_cmd = cmd
+        elif mode == 'singularity':
+            modded_cmd = f'singularity run --writable {container_uri} {cmd}'
+        else:
+            raise NameError("Invalid container launch mode.")
+
+        try:
+            proc = subprocess.Popen(modded_cmd,
+                                    stdout=subprocess.PIPE,
+                                    stderr=subprocess.PIPE,
+                                    shell=True)
+
+        except Exception as e:
+            logger.error("Got an error in worker launch, got error {}".format(e))
+            raise
+
+        return proc
+
+    def get_next_worker_q(self, new_worker_map):
+        """ Helper function to generate a queue of next workers to spin up .
+            From a mapping generated by the scheduler
+
+                Parameters
+                ----------
+                new_worker_map : dict
+                   {worker_type: total_number_of_containers,...}
+
+                Returns
+                ---------
+                Queue containing the next workers the system should spin-up.
+        """
+
+        next_worker_q = Queue()
```

Comment:
I don't think we need a queue here. We should be able to substitute with a list and use.
list.pop() and len(list)

Reply:
List-queue issue fixed.